morpheus/stages/input/control_message_kafka_source_stage.py

```diff
@@ -0,0 +1,185 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import time
+import typing
+
+import confluent_kafka as ck
+import mrc
+import pandas as pd
+
+from morpheus.cli.register_stage import register_stage
+from morpheus.config import Config
+from morpheus.config import PipelineModes
+from morpheus.messages.message_control import MessageControl
+from morpheus.pipeline.preallocator_mixin import PreallocatorMixin
+from morpheus.pipeline.single_output_source import SingleOutputSource
+from morpheus.pipeline.stream_pair import StreamPair
+from morpheus.stages.input.kafka_source_stage import AutoOffsetReset
+
+logger = logging.getLogger(__name__)
+
+
+@register_stage("from-cm-kafka", modes=[PipelineModes.AE])
+class ControlMessageKafkaSourceStage(PreallocatorMixin, SingleOutputSource):
```

Comment:
It would be much better to reimplement this or `KafkaSourceStage` to reuse the same class so code is not duplicated

Reply:
@bsuryadevara Can you refresh me on the details here; I know we'd talked about this at some length, but there was a reason that it didn't make sense. Or am I miss-remembering?

Reply:
There are multiple reasons why we needed to write different implementations:

Current DFP modules implementation pipeline is running with the use_cpp mode set to true. When we use the existing KafkaSourceStage, it ends up using the c++ implementation. This implementation uses `cudf.io.json` which is split the control message into multiple columns.

We attempted to use use_cpp=False mode, but unfortunately, we don't have the option with the existing Python implementation to choose the engine type or pass extra arguments.

KafkaSourceStage returns MessageMeta with a batch of messages. However, in this case, the ControlMessage  itself is a metadata that instructs how and where to load the data, as well as what tasks need to be performed. In this scenario, we need to restrict set the pipeline batch size to 1 by all means if above works.

Reply:
@bsuryadevara Perfect, thanks.